### DISCUSSION TITLE
`number | null` represented as `Character varying`!

### DISCUSSION THREAD

**Discussion by corrm** [mikro-orm/mikro-orm#6037] [Q&A] [answered]
I use postgraSQL

in my User entity: (there 26 field before it)
```ts
@Property({ nullable: true })
public lastDailyRewardClaimTime: number | null;
```

![image](https://github.com/user-attachments/assets/d63b964e-c02c-4e12-9c6a-d78cee356525)
But when i remove `| null` from the type declaration represented as a normal integer.

There is another problem i discovered:
When I add `type: "bigint"`
```ts
@Property({ type: "bigint", nullable: true })
public lastDailyRewardClaimTime: number | null;
```
and print that:
```ts
console.log(typeof user.lastDailyRewardClaimTime);
console.log(user.lastDailyRewardClaimTime);

// bigint
// 1726618503658n
```
BUT when i change field to:
```ts
@Property({ type: "bigint", nullable: true })
public lastDailyRewardClaimTime: number;

// number
// 1726618620645
```

For me, type should be `number` even with `number | null` as a type.
Even if I use `columnType` instead of `type` gives the same result.

_Thanks for your time again again_ :smile:

**Comment by B4nan** [comment]
Yes, that is all correct. That's how reflect-metadata works, it won't infer any union type.

**Reply by corrm** [reply]
Even if I use `columnType` instead of type gives the same result. so should i use both of them? `columnType` and `type`?

**Reply by B4nan** [reply, marked as answer]
Runtime type is inferred from your TS type via reflect metadata. If you use bigint column, it can be still mapped to either number, string or bigint on runtime, and that works based on the runtime type. So if you use a union type as in `number | null`, we can't infer anything and we fallback to the default. If you have a simple type, reflect metadata works, and the ORM will respect that.

https://mikro-orm.io/docs/using-bigint-pks

In that page, the examples are all explicit, but it would work the same without the explicit type instance:

```ts
@PrimaryKey({ type: 'bigint' })
id1: bigint;

@PrimaryKey({ type: 'bigint' })
id2: string;

@PrimaryKey({ type: 'bigint' })
id3: number;
```

But it wouldn't work if you use union types, there you would have to use it explicitly, e.g. `@PrimaryKey({ type: new BigIntType('number') })`. You can also use `runtimeType` option to set this explicitly.

**Reply by corrm** [reply]
I think i will stick to this syntax:
```ts
@Property({ type: "bigint", nullable: true })
public lastDailyRewardClaimTime?: number;
```

And deal with `undefined` and `null`(when data retrieved from DB `undefined` represented as `null`).

_Thanks_

**Reply by corrm** [reply]
@B4nan there is better way to handle that:
> And deal with `undefined` and `null`(when data retrieved from DB `undefined` represented as `null`).

The current behavior `{ nullable: true }` sets the field(even marked with `?`) to `null` instead of `undefined`?
